File: aoc2024/day01/day01.py
from collections import Counter


def parse(buffer: str) -> (list[int], list[int]):
    lines = [words for words in buffer.split("\n")]
    list_of_pairs = [[int(w) for w in line.split()] for line in lines if len(line) > 0]
    return list(zip(*list_of_pairs))


def part01(filename: str):
    with open(filename) as f:
        buffer = f.read()
    left, right = parse(buffer)
    left = sorted(left)
    right = sorted(right)
    return sum([abs(l - r) for l, r in zip(left, right)])


def similarity_score(left: list[int], right: list[int]):
    left_count = Counter(left)
    right_count = Counter(right)
    components = [
        left_count[value] * value * right_count[value] for value in left_count
    ]
    return sum(components)


def part02(filename: str):
    with open(filename) as f:
        buffer = f.read()
    left, right = parse(buffer)
    return similarity_score(left, right)


def example():
    print(part01("example1.txt"))
    print(part02("example1.txt"))


def part01_our_input():
    print(part01("input.txt"))


def part02_our_input():
    print(part02("input.txt"))


if __name__ == "__main__":
    example()
    part01_our_input()
    part02_our_input()

File: aoc2024/day01/day01.hs
import Data.List (sort)
import qualified Data.Map as M


parse buffer = map words $ lines buffer

parseInt :: String -> Int
parseInt = read

changeType :: [[String]] -> [[Int]]
changeType = ((fmap . fmap) parseInt)

transpose :: [[a]] -> [[a]]
transpose ([]:_) = [] 
transpose lol = (map head lol):(transpose (map tail lol))

twoLists :: String -> [[Int]] 
twoLists buffer = transpose $ changeType $ parse buffer

manhattan :: [[Int]] -> Int
manhattan xs = sum $ map (\x -> abs(x!!0 - x!!1)) $ transpose $ sort <$> xs

makeCounter :: [Int] -> M.Map Int Int
makeCounter xs = M.fromListWith (+) $ map (\x -> (x,1)) xs

similarity :: [Int] -> [Int] -> Int
similarity left right = M.foldrWithKey keyReducer 0 (makeCounter left)
  where rc = makeCounter right
        keyReducer n freqLeft acc = acc + n * freqLeft * (M.findWithDefault 0 n rc)

part1 :: String -> IO ()
part1 filename = do
		    myLists <- twoLists <$> readFile filename
		    let answer = manhattan myLists
		    print(answer)

part2 :: String -> IO ()
part2 filename = do
		    (left:right:_) <- twoLists <$> readFile filename
		    let answer = similarity left right
		    print(answer
